src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -0,0 +1,146 @@
+  if (io_service_thread_->joinable()) {
+    io_service_thread_->join();
+  }
+}
+
+absl::optional<rpc::TaskEvents>
+GcsTaskManager::GcsTaskManagerStorage::AddOrReplaceTaskEvent(
+    rpc::TaskEvents events_by_task) {
+  RAY_LOG_EVERY_MS(INFO, 10000)
+      << "GcsTaskManagerStorage currently stores " << task_events_.size()
+      << " task event entries, approximate size="
+      << 1.0 * num_bytes_task_events_ / 1024 / 1024 << "MiB";
+
+  TaskID task_id = TaskID::FromBinary(events_by_task.task_id());
+  int32_t attempt_number = events_by_task.attempt_number();
+  TaskAttempt task_attempt = std::make_pair<>(task_id, attempt_number);
+
+  // GCS perform merging of events/updates for a single task attempt from multiple
+  // reports.
+  auto itr = task_attempt_index_.find(task_attempt);
+  if (itr != task_attempt_index_.end()) {
+    // Existing task attempt entry, merge.
+    auto idx = itr->second;
+    auto &existing_events = task_events_.at(idx);
+
+    // Update the events.
+    num_bytes_task_events_ -= existing_events.ByteSizeLong();
+    existing_events.MergeFrom(events_by_task);
+    num_bytes_task_events_ += existing_events.ByteSizeLong();
+
+    return absl::nullopt;
+  }
+
+  // A new task event, add to storage and index.
+
+  // If limit enforced, replace one.
+  if (max_num_task_events_ > 0 && task_events_.size() >= max_num_task_events_) {
+    RAY_LOG_EVERY_MS(WARNING, 10000)
+        << "Max number of tasks event (" << max_num_task_events_
+        << ") allowed is reached. Old task events will be overwritten. Set "
+           "`RAY_task_events_max_num_task_in_gcs` to a higher value to "
+           "store more.";
+
+    num_bytes_task_events_ -= task_events_[next_idx_to_overwrite_].ByteSizeLong();
+    num_bytes_task_events_ += events_by_task.ByteSizeLong();
+
+    // Change the underlying storage.
+    auto &to_replaced = task_events_.at(next_idx_to_overwrite_);
+    std::swap(to_replaced, events_by_task);
+    auto replaced = events_by_task;
+
+    // Update index.
+    TaskAttempt replaced_attempt = std::make_pair<>(
+        TaskID::FromBinary(replaced.task_id()), replaced.attempt_number());
+    task_attempt_index_.erase(replaced_attempt);
+    task_attempt_index_[task_attempt] = next_idx_to_overwrite_;
+
+    // Update iter.
+    next_idx_to_overwrite_ = (next_idx_to_overwrite_ + 1) % max_num_task_events_;
+
+    return replaced;
+  }
+
+  // Add to index.
+  task_attempt_index_[task_attempt] = task_events_.size();
+  // Add a new task events.
+  task_events_.push_back(std::move(events_by_task));
+  return absl::nullopt;
+}
+
+void GcsTaskManager::HandleAddTaskEventData(rpc::AddTaskEventDataRequest request,
+                                            rpc::AddTaskEventDataReply *reply,
+                                            rpc::SendReplyCallback send_reply_callback) {
+  RAY_LOG(DEBUG) << "Adding task state event:" << request.data().ShortDebugString();
+  // Dispatch to the handler
+  io_service_.post(
```

Comment:
Is this necessary? I think this method should be dispatched to this io service automatically?